ambari-server/src/main/resources/stack-hooks/before-SET_KEYTAB/scripts/hook.py

```diff
@@ -0,0 +1,38 @@
+"""
+Licensed to the Apache Software Foundation (ASF) under one
+or more contributor license agreements.  See the NOTICE file
+distributed with this work for additional information
+regarding copyright ownership.  The ASF licenses this file
+to you under the Apache License, Version 2.0 (the
+"License"); you may not use this file except in compliance
+with the License.  You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+
+"""
+
+from resource_management import *
+
+class BeforeSetKeytabHook(Hook):
```

Comment:
Here's the deal on this one. The keytab generation failed b/c of new users added in the target stack. Because we don't use normal `INSTALL` commands to distribute the new stack, the `before-INSTALL` hook was never called and those users were never created.

The `before-ANY` hook has the user/group creation logic tightly coupled inside of it, so we needed a simple, obtrusive way to run this code (inside of the hook) before the keytab command ran. This was the least invasive way of doing it.